File: protocol-autoresearch/experiments/tier2_000_framework_integrator_proof_envelope_content_type/sdk/src/envelope.ts
import { z } from 'zod';

/**
 * MIME content type for Bolyra proof envelopes.
 */
export const CONTENT_TYPE = 'application/bolyra-proof+json' as const;

/** Current envelope schema version. */
export const ENVELOPE_VERSION = '1.0' as const;

/** Proof data matching snarkjs Groth16/PLONK output. */
export interface ProofData {
  pi_a: string[];
  pi_b: string[][];
  pi_c: string[];
  protocol: 'groth16' | 'plonk';
}

/** Metadata attached to every proof envelope. */
export interface ProofMetadata {
  prover: string;
  timestamp: string; // ISO 8601
}

/** Canonical proof envelope for HTTP transport. */
export interface ProofEnvelope {
  version: string;
  circuit: string;
  publicSignals: string[];
  proof: ProofData;
  metadata: ProofMetadata;
}

// ── Zod schemas ──────────────────────────────────────────────────────────────

const ProofDataSchema = z.object({
  pi_a: z.array(z.string()),
  pi_b: z.array(z.array(z.string())),
  pi_c: z.array(z.string()),
  protocol: z.enum(['groth16', 'plonk']),
});

const ProofMetadataSchema = z.object({
  prover: z.string().min(1),
  timestamp: z.string().datetime(),
});

export const ProofEnvelopeSchema = z.object({
  version: z.string(),
  circuit: z.string().min(1),
  publicSignals: z.array(z.string()),
  proof: ProofDataSchema,
  metadata: ProofMetadataSchema,
});

// ── Version helpers ──────────────────────────────────────────────────────────

function parseMajorVersion(version: string): number {
  const major = parseInt(version.split('.')[0], 10);
  if (isNaN(major)) throw new Error(`Invalid version string: ${version}`);
  return major;
}

function assertSupportedVersion(version: string): void {
  const major = parseMajorVersion(version);
  const supportedMajor = parseMajorVersion(ENVELOPE_VERSION);
  if (major !== supportedMajor) {
    throw new Error(
      `Unsupported envelope major version ${major}; expected ${supportedMajor}`
    );
  }
}

// ── Public API ───────────────────────────────────────────────────────────────

/**
 * Serialize a ProofEnvelope to a JSON string.
 */
export function serializeEnvelope(envelope: ProofEnvelope): string {
  ProofEnvelopeSchema.parse(envelope);
  return JSON.stringify(envelope);
}

/**
 * Deserialize a JSON string into a validated ProofEnvelope.
 * Throws on validation failure or unsupported major version.
 */
export function deserializeEnvelope(raw: string): ProofEnvelope {
  const parsed = JSON.parse(raw);
  const envelope = ProofEnvelopeSchema.parse(parsed);
  assertSupportedVersion(envelope.version);
  return envelope;
}

/**
 * Validate an unknown object as a ProofEnvelope.
 * Returns the validated envelope or throws a ZodError.
 */
export function validateEnvelope(data: unknown): ProofEnvelope {
  const envelope = ProofEnvelopeSchema.parse(data);
  assertSupportedVersion(envelope.version);
  return envelope;
}

/**
 * Wrap raw snarkjs proof output into a ProofEnvelope.
 */
export function envelopeFromSnarkjsProof(
  circuit: string,
  proof: { pi_a: string[]; pi_b: string[][]; pi_c: string[]; protocol: string },
  publicSignals: string[],
  prover: string = '@bolyra/sdk'
): ProofEnvelope {
  return {
    version: ENVELOPE_VERSION,
    circuit,
    publicSignals,
    proof: {
      pi_a: proof.pi_a,
      pi_b: proof.pi_b,
      pi_c: proof.pi_c,
      protocol: proof.protocol as 'groth16' | 'plonk',
    },
    metadata: {
      prover,
      timestamp: new Date().toISOString(),
    },
  };
}

File: protocol-autoresearch/experiments/tier2_000_framework_integrator_proof_envelope_content_type/sdk-python/bolyra/envelope.py
"""Bolyra proof envelope — canonical wire format for HTTP transport.

Mirrors the TypeScript SDK's envelope.ts schema exactly.
"""

from __future__ import annotations

import json
from datetime import datetime, timezone
from typing import List, Literal

from pydantic import BaseModel, field_validator

CONTENT_TYPE: str = "application/bolyra-proof+json"
ENVELOPE_VERSION: str = "1.0"

_SUPPORTED_MAJOR = int(ENVELOPE_VERSION.split(".")[0])


class ProofData(BaseModel):
    """Proof fields matching snarkjs Groth16/PLONK output."""

    pi_a: List[str]
    pi_b: List[List[str]]
    pi_c: List[str]
    protocol: Literal["groth16", "plonk"]


class ProofMetadata(BaseModel):
    """Metadata attached to every proof envelope."""

    prover: str
    timestamp: str  # ISO 8601

    @field_validator("prover")
    @classmethod
    def prover_not_empty(cls, v: str) -> str:
        if not v:
            raise ValueError("prover must not be empty")
        return v


class ProofEnvelope(BaseModel):
    """Canonical proof envelope for HTTP transport."""

    version: str
    circuit: str
    publicSignals: List[str]  # noqa: N815 — matches wire format
    proof: ProofData
    metadata: ProofMetadata

    @field_validator("circuit")
    @classmethod
    def circuit_not_empty(cls, v: str) -> str:
        if not v:
            raise ValueError("circuit must not be empty")
        return v

    def to_json(self) -> str:
        """Serialize to canonical JSON string."""
        return self.model_dump_json()

    @classmethod
    def from_json(cls, raw: str) -> "ProofEnvelope":
        """Deserialize and validate from JSON string.

        Raises ValueError on unsupported major version.
        """
        envelope = cls.model_validate_json(raw)
        _assert_supported_version(envelope.version)
        return envelope


def _parse_major_version(version: str) -> int:
    try:
        return int(version.split(".")[0])
    except (ValueError, IndexError) as exc:
        raise ValueError(f"Invalid version string: {version}") from exc


def _assert_supported_version(version: str) -> None:
    major = _parse_major_version(version)
    if major != _SUPPORTED_MAJOR:
        raise ValueError(
            f"Unsupported envelope major version {major}; expected {_SUPPORTED_MAJOR}"
        )


def envelope_from_snarkjs_proof(
    circuit: str,
    proof: dict,
    public_signals: List[str],
    prover: str = "bolyra-python",
) -> ProofEnvelope:
    """Wrap raw snarkjs proof output into a ProofEnvelope."""
    return ProofEnvelope(
        version=ENVELOPE_VERSION,
        circuit=circuit,
        publicSignals=public_signals,
        proof=ProofData(
            pi_a=proof["pi_a"],
            pi_b=proof["pi_b"],
            pi_c=proof["pi_c"],
            protocol=proof["protocol"],
        ),
        metadata=ProofMetadata(
            prover=prover,
            timestamp=datetime.now(timezone.utc).isoformat(),
        ),
    )
